### ISSUE TITLE
Unix start scripts unexpectedly cd to application dir when started from ~on OS X.

### ISSUE THREAD

**Issue by robertpanzer** [gradle/gradle#5978]
As found in https://github.com/asciidoctor/asciidoctorj/issues/662 It looks like since Gradle 3.0 the start script generated by the application plugin unexpectedly changes the directory to the application directory when started from ~ on OS X.

For an application that receives file names as arguments it is impossible to find out that files in ~ are meant.

This seems to be a regression introduced with https://github.com/gradle/gradle/commit/a2b76dff7d879e6f0997f9a4fe746216e05011a6

### Expected Behavior
The script should not change dir.

### Current Behavior
The start script changes the worked directory.

### Context
For AsciidoctorJ it makes the whole application fail when trying to render a document in ~ while it works as expected everywhere else.

### Steps to Reproduce (for bugs)
Download http://jcenter.bintray.com/org/asciidoctor/asciidoctorj/1.5.7/:asciidoctorj-1.5.7-bin.zip und unpack it.
Then create a file `~/test.adoc` with this content:
```
= Test

Test
```

Then execute:
```
$ cd
$ ~/Downloads/asciidoctorj-1.5.7/bin/asciidoctorj test.adoc
asciidoctor: FAILED: input file(s) '/Users/robertpanzer/Downloads/asciidoctorj-1.5.7/bin/test.adoc' missing or cannot be read
Exception in thread "main" java.lang.IllegalArgumentException: asciidoctor: FAILED: input file(s) '/Users/robertpanzer/Downloads/asciidoctorj-1.5.7/bin/test.adoc' missing or cannot be read
        at org.asciidoctor.cli.AsciidoctorInvoker.renderInput(AsciidoctorInvoker.java:138)
        at org.asciidoctor.cli.AsciidoctorInvoker.invoke(AsciidoctorInvoker.java:65)
        at org.asciidoctor.cli.AsciidoctorInvoker.main(AsciidoctorInvoker.java:183)
```

### Your Environment

Unrelated to build, this is a problem of the generated script.

**Comment by gblues** [comment]
I just ran into this too. What a pain in the ass. Thanks @robertpanzer  for reporting this.

Robert's report is a little confusing because the location of files (both the application and whatever file path being passed on the command line) are irrelevant.

**What matters is whether or not you are in your home directory when you execute the auto-generated run script**

```
gradle -version

------------------------------------------------------------
Gradle 5.0
------------------------------------------------------------

Build time:   2018-11-26 11:48:43 UTC
Revision:     7fc6e5abf2fc5fe0824aec8a0f5462664dbcd987

Kotlin DSL:   1.0.4
Kotlin:       1.3.10
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.13 compiled on July 10 2018
JVM:          11.0.1 (Oracle Corporation 11.0.1+13)
OS:           Mac OS X 10.14.2 x86_64
```

I've attached a very simple project to demonstrate:
[gradlebug_demo.zip](https://github.com/gradle/gradle/files/2797849/gradlebug_demo.zip)

This projected was created via `gradle init` and the only change from the boilerplate code is to
make the output be the working directory instead of "Hello, world."

Note the difference in output depending on where you start it from:

```lang=bash
# This test performed on Mac OSX Mojave 10.14.2
# start in the home directory
➜  ~ cd tmp
➜  tmp ./gradlebug_demo/bin/gradlebug_demo
# When run from ~/tmp, the output is as you expect.
/Users/nstrong/tmp/.
# Now go back to the home directory
➜  tmp cd
# run the same application from the home directory
➜  ~ ./tmp/gradlebug_demo/bin/gradlebug_demo
# the working directory is now the location of the generated gradlebug_demo script, not my home
# directory as expected
/Users/nstrong/tmp/gradlebug_demo/bin/.
# running the application via an absolute path makes no difference
➜  ~ $(pwd)/tmp/gradlebug_demo/bin/gradlebug_demo
/Users/nstrong/tmp/gradlebug_demo/bin/.
```

**Comment by big-guy** [comment]
Hmm, so this comes from https://github.com/gradle/gradle/pull/605

I think the intention there was to fix it so `gradlew` could be run from the macOS Finder and the directory of the script == project directory.

Since the same infrastructure is used to generate the application scripts too, the `application` plugin inherited the same quirk. 

Would either of you (@gblues or @robertpanzer) be interested in contributing a fix?  It seems like we should limit this to only the gradlew script and maybe make the check more specific (macOS, CWD==HOME, $0==absolute path).

**Comment by robertpanzer** [comment]
I am fine, I don't need a fix for this bug anymore.
I circumvented this problem by providing my own script template.

What is the use case for calling gradlew from Finder?
You will always need additional parameters.

**Comment by gblues** [comment]
launching GUI Java apps, I expect, @robertpanzer.

I've got some cycles today, I'll take a look at it to see if I can wrap my brain around it quickly.

**Comment by big-guy** [comment]
> What is the use case for calling gradlew from Finder?

I agree that it's not very useful, but if you had default tasks defined for the build, it would let you double click on gradlew and do something. 

> I've got some cycles today, I'll take a look at it to see if I can wrap my brain around it quickly.

Thanks @gblues

**Comment by gblues** [comment]
Note that if someone else knows a reliable way to detect "run from finder" that isn't an inference (e.g. assuming based on pwd or by whether or not $0 is an absolute path), I'm all for it. I wasn't able to find an obvious one.

**Comment by gblues** [comment]
@big-guy I've submitted a PR with the recommended fix.

**Comment by Calvin-L** [comment]
I think this issue can be closed. It looks like it was fixed in PR #10107.

**Comment by bmuskalla** [comment]
Thanks @Calvin-L
